app/lib/account/backend.dart

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:googleapis/oauth2/v2.dart' as oauth2_v2;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:retry/retry.dart';
+import 'package:uuid/uuid.dart';
+
+import '../frontend/models.dart' show PackageVersion;
+
+import 'models.dart';
+
+final _logger = new Logger('pub.account.backend');
+final _retry = RetryOptions();
+final _uuid = Uuid();
+
+/// The pub client's OAuth2 identifier.
+final _pubAudience = '818368855108-8grd2eg9tj9f38os6f1urbcvsq399u8n.apps.'
+    'googleusercontent.com';
+
+/// Sets the account backend service.
+void registerAccountBackend(AccountBackend backend) =>
+    ss.register(#_accountBackend, backend);
+
+/// The active account backend service.
+AccountBackend get accountBackend =>
+    ss.lookup(#_accountBackend) as AccountBackend;
+
+/// Represents the backend for the account handling and authentication.
+class AccountBackend {
+  final DatastoreDB _db;
+  final _defaultAuthProvider = GoogleOauth2AuthProvider(_pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+  }
+
+  /// Returns the `User` entry for the [userId] or null if it does not exists.
+  Future<User> lookupUserById(String userId) async {
+    final key = _db.emptyKey.append(User, id: userId);
+    return (await _db.lookup<User>([key])).single;
+  }
+
+  /// Returns the `User` entry that is associated with the [accessToken].
+  ///
+  /// When no entry exists in Datastore, this method will create a new one.
+  ///
+  /// When the authenticated e-mail of the user changes, the email field will
+  /// be updated to the latest one.
+  ///
+  /// The method returns null if the access token is invalid.
+  Future<User> authenticateWithAccessToken(String accessToken) async {
+    final auth = await _defaultAuthProvider.tryAuthenticate(accessToken);
+    if (auth == null) {
+      return null;
+    }
+    return await _lookupOrCreateUserByOauthUserId(auth);
+  }
+
+  Future<User> _lookupOrCreateUserByOauthUserId(AuthResult auth) async {
+    final mappingKey = _db.emptyKey.append(User, id: auth.userId);
+
+    final user = await _retry.retry(() async {
+      // Check existing mapping.
+      final mapping = (await _db.lookup<OAuthUserID>([mappingKey])).single;
+      if (mapping != null) {
+        final user = (await _db.lookup<User>([mapping.userIdKey])).single;
+        // TODO: we should probably have some kind of consistency mitigation
+        if (user == null) {
+          throw StateError('Incomplete OAuth userId mapping.');
```

Comment:
Throw an `AssertionError` instead.. and make the message highlight a datastore invariant violation.